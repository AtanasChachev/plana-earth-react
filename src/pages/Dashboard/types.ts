import { Product } from 'models/products';
import { ProductStatisticsValue } from 'models/products';

export interface DashboardProps {
  products: Product[];
}

export interface UseDashboardProps {
  products: Product[];
}

export interface Return {
  isFilterActive: boolean;
  products: Product[];
  currentProduct: Product | null;
  tableData: ProductStatisticsValue[];
  isLoading?: boolean;
  handleProductChange: (value: string) => void;
  handleCountryChange: (value: string, countryName?: string) => void;
  handleDatepickerStartDateChange: (date: string) => void;
  handleDatepickerEndDateChange: (date: string) => void;
  renderChartTitle: () => string;
  handleChartFilterButtonsChange: (intervalValue: string) => void;
}
