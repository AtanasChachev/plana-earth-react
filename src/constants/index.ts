export * from './chart';
export * from './countries';
export * from './theme';
export * from './toast';