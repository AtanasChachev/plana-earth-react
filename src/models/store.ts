import { ChartState } from './chart';
import { ProductState } from './products';

export type State = {
  productsState: ProductState;
  chartState: ChartState;
};

export type StateAction = {
  type: symbol;
  payload: never;
};