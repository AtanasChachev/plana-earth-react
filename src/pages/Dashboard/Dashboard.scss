@import 'styles/utils/mixins';

.dashboard {
  @include breakpoint(min, width, 960px) {
    height: 100vh;
    overflow: hidden;
  }
}