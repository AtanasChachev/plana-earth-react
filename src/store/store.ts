import { createStore, combineReducers } from 'redux';
import ProductsReducer from './reducers/products';

const reducers = combineReducers({
  products: ProductsReducer,
});

export const store = createStore(reducers);
