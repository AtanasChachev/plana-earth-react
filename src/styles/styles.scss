@import './config/variables';
@import './utils/mixins';