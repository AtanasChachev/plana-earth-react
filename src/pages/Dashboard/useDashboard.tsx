import { useEffect, useState } from 'react';
import { ActiveProductFilters,  ProductStatisticsValue } from 'models/products';
import { capitalizeWords, formatDate } from 'utils/helpers';
import { Product } from 'models/products';
import { DEFAULT_ACTIVE_FILTERS } from 'constants/products';
import { useGetProductStatistic } from 'services/products/getProductStatistics';
import { getObjectValues } from 'utils/helpers';

interface UseDashboardProps {
  products: Product[];
}

interface Return {
  isFilterActive: boolean;
  products: Product[];
  currentProduct: Product | null;
  tableData: ProductStatisticsValue[];
  isLoading?: boolean;
  handleProductChange: (value: string) => void;
  handleCountryChange: (value: string, countryName?: string) => void;
  handleDatepickerStartDateChange: (date: string) => void;
  handleDatepickerEndDateChange: (date: string) => void;
  renderChartTitle: () => string;
  handleChartFilterButtonsChange: (intervalValue: string) => void;
}

export const useDashboard = ({ products }: UseDashboardProps): Return => {
  const [currentProduct, updateCurrentProduct] = useState<Product | null>(null);
  const [activeFilters, updateActiveFilters] = useState<ActiveProductFilters>({ ...DEFAULT_ACTIVE_FILTERS });

  const { productsRangeData, isFetching: isLoading }  = useGetProductStatistic({
    name: activeFilters.name,
    id: activeFilters.country.id,
    startDate: activeFilters.startDate,
    endDate: activeFilters.endDate,
    interval: activeFilters.interval,
  });

  const tableData = productsRangeData.reduce((mappedData: ProductStatisticsValue[], currentItem) => {
    mappedData.push({
      time: formatDate(currentItem.time.interval_start, 'YYYY-MM-DD'),
      ...currentItem.value,
    }); 

    return mappedData;
  }, []);

  const [isFilterActive, updateFilterActive] = useState<boolean>(false);

  const handleAllFiltersUpdateEffect = () => {
    const allFiltersActive = getObjectValues(activeFilters).every(obj => obj.length);
    updateFilterActive(allFiltersActive);
  };

  useEffect(handleAllFiltersUpdateEffect, [activeFilters]);

  const renderChartTitle = (): string => isFilterActive ? 
    `Showing data for: ${capitalizeWords(activeFilters.name)} in ${activeFilters.country.name}` : '';

  const handleProductChange = (value: string) => {
    const option = products.find(product => product.id === value);

    handleUpdateActiveFilter('name', option?.name || '');
    updateCurrentProduct(option as Product);
  };

  const handleCountryChange = (value: string, countryName?: string) => {
    handleUpdateActiveFilter('country', { 
      id: value, 
      name: countryName || '', 
    });
  };

  const handleDatepickerStartDateChange = (date: string) => handleUpdateActiveFilter('startDate', date);
  const handleDatepickerEndDateChange = (date: string) => handleUpdateActiveFilter('endDate', date);
  const handleChartFilterButtonsChange = (intervalValue: string) => handleUpdateActiveFilter('interval', intervalValue);

  const handleUpdateActiveFilter = (prop: string, value: string | ActiveProductFilters['country']) => 
    updateActiveFilters(object => ({
      ...object,
      [prop]: value,
    }));

  return {
    isFilterActive,
    products,
    tableData,
    currentProduct,
    isLoading,
    handleProductChange,
    handleCountryChange,
    handleDatepickerStartDateChange,
    handleDatepickerEndDateChange,
    renderChartTitle,
    handleChartFilterButtonsChange,
  };
};