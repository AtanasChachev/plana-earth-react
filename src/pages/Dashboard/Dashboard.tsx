import Grid from '@material-ui/core/Grid';
import { Section, SelectComponent, Datepicker, Chart, ChartFilterButtons, EmptyResults } from 'components/index';
import { COUNTRIES } from 'constants/index';
import { useDashboard } from './useDashboard';

import './Dashboard.scss';

const Dashboard = (): JSX.Element => {
  const {
    isFilterActive,
    products,
    currentProduct,
    tableData,
    handleProductChange,
    handleCountryChange,
    handleDatepickerStartDateChange,
    handleDatepickerEndDateChange,
    renderChartTitle,
    handleChartFilterButtonsChange,
  } = useDashboard();

  return (
    <Grid container className="dashboard"> 
      <Grid item xs={12} md={4} lg={3}>
        <Section
          isEmptyBlockActive={false} 
          isFullHeight={true}
          shAlignInlineBlockOnMobile={true}
          headerTitle='Product filters:'>
            {
              !!products.length ? <SelectComponent 
                ariaLabel="Select dropdown - choose product"
                className="dashboard__form__field"
                options={products}
                placeholder={'Product'}
                onChange={handleProductChange} /> : <></>
            }

          <SelectComponent 
            ariaLabel="Select dropdown - choose country"
            className={`dashboard__form__field dashboard__form__field--animated ${currentProduct ? 'dashboard__form__field--animated-visible' : ''}`}
            options={COUNTRIES}
            placeholder={'Country'}
            onChange={handleCountryChange} />

          <Datepicker 
            ariaLabel="Datepicker - choose the start date for the filter"
            className={`dashboard__form__field dashboard__form__field--animated ${currentProduct ? 'dashboard__form__field--animated-visible' : ''}`}
            label='Start Date'
            minDate={currentProduct ? currentProduct.first : ''}
            maxDate={currentProduct ? currentProduct.last : ''}
            onChange={handleDatepickerStartDateChange}
          />

          <Datepicker  
            ariaLabel="Datepicker - choose the end date for the filter"
            className={`dashboard__form__field dashboard__form__field--animated ${currentProduct ? 'dashboard__form__field--animated-visible' : ''}`}
            label='End Date'
            minDate={currentProduct ? currentProduct.first : ''}
            maxDate={currentProduct ? currentProduct.last : ''}
            onChange={handleDatepickerEndDateChange}
          />
        </Section>
      </Grid>

      <Grid item xs={12} md={8} lg={9}>
        <Section 
          isEmptyBlockActive={!isFilterActive}
          headerTitle={renderChartTitle()}
          shRenderLoader={true}
          isFullHeight={true}>
            {
              isFilterActive ? <>
                <ChartFilterButtons onClick={handleChartFilterButtonsChange} />
                <Chart data={tableData} />
              </> : <EmptyResults />
            }
        </Section>
      </Grid>
    </Grid>
  );
};

export { Dashboard };