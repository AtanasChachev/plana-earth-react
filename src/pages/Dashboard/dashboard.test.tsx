import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from 'store/store';
import { Dashboard } from './Dashboard';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';
import { MuiThemeProvider, createTheme } from '@material-ui/core/styles';
import { THEME } from 'constants/index';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

const theme = createTheme({
  palette: {
    primary: {
      main: THEME.palette.primary.main,
      contrastText: THEME.palette.primary.contrastText,
    },
  },
});

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: false,
    },
  },
});

const mockProducts = [
  {
    'description': 'Dry-air mixing ratio of methane for cloud-free observations with a spatial resolution of 7x7km2 observed at about 13:30 local solar time from spectra measured by TROPOMI, total column',
    'name': 'methane',
    'product_variable': 'methane_mixing_ratio_bias_corrected',
    'id': 'methane',
    'first': '2019-01-01T00:26:47.791000Z',
    'last': '2023-03-03T03:16:44.542000Z',
  },
  {
    'description': 'Atmospheric content of carbon monoxide in `mol m¯²`, total column',
    'name': 'carbonmonoxide',
    'product_variable': 'carbonmonoxide_total_column',
    'id': 'carbonmonoxide',
    'first': '2019-01-01T00:23:31.235000Z',
    'last': '2023-03-03T13:28:56.552000Z',
  },
  {
    'description': 'Atmospheric content of ozone in `mol m¯²`, total column',
    'name': 'ozone',
    'product_variable': 'ozone_total_vertical_column',
    'id': 'ozone',
    'first': '2019-04-23T17:29:01.796000Z',
    'last': '2023-03-03T08:26:16.163000Z',
  },
  {
    'description': 'Nitrogen dioxide tropospheric column with a spatial resolution of 7x3.5km2 observed at about 13:30 local solar time from spectra measured by TROPOMI, total column',
    'name': 'nitrogendioxide',
    'product_variable': 'nitrogendioxide_tropospheric_column',
    'id': 'nitrogendioxide',
    'first': '2019-01-01T01:22:50.197000Z',
    'last': '2023-02-24T22:27:56.663000Z',
  },
];

const setup = () => {
  const { asFragment } = render(
    <QueryClientProvider client={queryClient}>
      <Provider store={store}>
        <MuiPickersUtilsProvider utils={MomentUtils}>
          <MuiThemeProvider theme={theme}>
            <Dashboard products={mockProducts} />
          </MuiThemeProvider>
        </MuiPickersUtilsProvider>,
      </Provider>
    </QueryClientProvider>,
  );

  return { asFragment };
};

describe('<Dashboard />', () => {
  test('<Dashboard /> should render correctly', () => {
    const { asFragment } = setup();

    expect(asFragment()).toMatchSnapshot();
  });
});