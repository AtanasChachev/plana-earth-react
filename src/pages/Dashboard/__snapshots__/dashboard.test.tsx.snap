// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> <Dashboard /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root dashboard MuiGrid-container"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 MuiGrid-grid-lg-3"
    >
      <div
        class="section"
      >
        <div
          class="section__inner section__inner--full-height"
        >
          <div
            class="section__inner__header"
          >
            <h3
              class="section__inner__header__title"
            >
              Product filters:
            </h3>
          </div>
          <div
            class="section__inner__content section__inner__content--mobile-inline-blocks "
          >
            0
            <div
              class="MuiFormControl-root select dashboard__form__field dashboard__form__field--animated "
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root select__label MuiInputLabel-formControl MuiInputLabel-animated"
                data-shrink="false"
              >
                Country
              </label>
              <div
                aria-label="Select dropdown - choose country"
                aria-required="true"
                class="MuiInputBase-root MuiInput-root MuiInput-underline select__form MuiInputBase-formControl MuiInput-formControl"
                data-testid="select"
              >
                <div
                  aria-haspopup="listbox"
                  class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                  role="button"
                  tabindex="0"
                >
                  <span>
                    â€‹
                  </span>
                </div>
                <input
                  aria-hidden="true"
                  class="MuiSelect-nativeInput"
                  tabindex="-1"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSelect-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </div>
            </div>
            <div
              aria-label="Datepicker - choose the start date for the filter"
              aria-required="true"
              class="MuiFormControl-root MuiTextField-root datepicker dashboard__form__field dashboard__form__field--animated  MuiFormControl-marginNormal"
              data-testid="datepicker"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                data-shrink="false"
              >
                Start Date
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
                  type="text"
                  value=""
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                >
                  <button
                    class="MuiButtonBase-root MuiIconButton-root"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                        />
                        <path
                          d="M0 0h24v24H0z"
                          fill="none"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
            </div>
            <div
              aria-label="Datepicker - choose the end date for the filter"
              aria-required="true"
              class="MuiFormControl-root MuiTextField-root datepicker dashboard__form__field dashboard__form__field--animated  MuiFormControl-marginNormal"
              data-testid="datepicker"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                data-shrink="false"
              >
                End Date
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
                  type="text"
                  value=""
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                >
                  <button
                    class="MuiButtonBase-root MuiIconButton-root"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                        />
                        <path
                          d="M0 0h24v24H0z"
                          fill="none"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8 MuiGrid-grid-lg-9"
    >
      <div
        class="section"
      >
        <div
          class="section__inner section__inner--full-height"
        >
          <div
            class="section__inner__content  section__inner__content--center"
          >
            <div
              class="empty-results"
            >
              <img
                alt="Search Icon"
                class="empty-results__image"
                src="search.png"
              />
              <h2
                class="empty-results__title"
              >
                No Results Found
              </h2>
              <p
                class="empty-results__description"
              >
                We were not able to find results by the given criteria. Please try to change the filters in the left column.
              </p>
            </div>
          </div>
          <div
            class="loader "
          >
            <div
              class="MuiCircularProgress-root loader__spinner MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
              role="progressbar"
              style="width: 60px; height: 60px;"
            >
              <svg
                class="MuiCircularProgress-svg"
                viewBox="22 22 44 44"
              >
                <circle
                  class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                  cx="44"
                  cy="44"
                  fill="none"
                  r="20.2"
                  stroke-width="3.6"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  ,
</DocumentFragment>
`;
